build_runner/lib/src/package_graph/package_graph.dart

```diff
@@ -147,8 +150,25 @@ class PackageGraph {
   /// last. For any two packages for which neither is a transitive dependency of
   /// the other the relative position of the packages within the cycle is
   /// non-deterministic.
+  @Deprecated('Use orderedPackageCycles instead')
   Iterable<PackageNode> get orderedPackages =>
-      _orderedPackages(root, new Set<PackageNode>());
+      orderedPackageCycles.expand((c) => c);
+
+  /// Collects packages into strongly connected subcomponents and orders those
+  /// components in reverse dependency order.
+  ///
+  /// A cycle is defined as a collection of packages where each package is
+  /// present in the transitive dependency graph of each other package.
+  ///
+  /// Cycles are ordered in reverse dependency order, so if package `A` depends
+  /// on package `B`, and they are not in a cycle, package `A` will exist in a
+  /// later sublist than `B`. Ordering within a cycle is non-deterministic,
```

Comment:
We should be able to make it deterministic pretty easily by just sorting alphabetically within cycles?

Reply:
I'm not sure it would be worthwhile to spend any time sorting. We can add that if we find a use case.

Reply:
The issue I am concerned about here is the ordering will actually have a real effect, so this makes builds non-deterministic.

Specifically if the packages try to read generated outputs from each other, for one of them `canRead` might return false. Which one that is depends on the cycle ordering.